### PR TITLE
feat(linter): eslint-plugin-react: no-direct-mutation-state

### PR DESCRIPTION
**Pull request by zhangrunzhao**
#1022 
try to implement `no-direct-mutation-state`

- code: https://github.com/jsx-eslint/eslint-plugin-react/blob/master/lib/rules/no-direct-mutation-state.js
- doc: https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/no-direct-mutation-state.md
- test: https://github.com/jsx-eslint/eslint-plugin-react/blob/master/tests/lib/rules/no-direct-mutation-state.js

### CODE REVIEW

**Comment by zhangrunzhao** [comment]
when i try to rebase main branch, throw a error.
wVcgn6sWXd

I could run normally before rebase, so I judged that it was not caused by me and initiated merge request

**Comment by camc314** [comment]
> when i try to rebase main branch, throw a error. wVcgn6sWXd
> 
> I could run normally before rebase, so I judged that it was not caused by me and initiated merge request

Hmm - says it's a network error? does hitting `https://index.crates.io` work in a browser?

File: crates/oxc_linter/src/rules/react/no_direct_mutation_state.rs [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+
+    is_constructor == 1 && is_call_expression_node == 0 || is_component == 0
+}
+
+impl Rule for NoDirectMutationState {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        match node.kind() {
+            AstKind::AssignmentExpression(assignment_expr) => {
+                if should_ignore_component(node, ctx) {
+                    return {};
```
Comment (camc314) [inline crates/oxc_linter/src/rules/react/no_direct_mutation_state.rs:177]: unneeded `{}`?

File: crates/oxc_linter/src/rules/react/no_direct_mutation_state.rs [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+        Expression::MemberExpression(member_expr) => match &member_expr.0 {
+            MemberExpression::StaticMemberExpression(expr) => Some(expr),
+            _ => None,
+        },
+        _ => None,
+    }
+}
+
+fn should_ignore_component<'a, 'b>(node: &'b AstNode<'a>, ctx: &'b LintContext<'a>) -> bool {
+    let mut is_constructor: i8 = 0;
```
Comment (camc314) [inline crates/oxc_linter/src/rules/react/no_direct_mutation_state.rs:149]: any reason why you are using an `i8` instead of a bool?

Comment (zhangrunzhao) [reply crates/oxc_linter/src/rules/react/no_direct_mutation_state.rs:149]: Hmm - using bool is a better solution, and I'll change that right away.

**Comment by zhangrunzhao** [comment]
> > when i try to rebase main branch, throw a error. wVcgn6sWXd
> > I could run normally before rebase, so I judged that it was not caused by me and initiated merge request
> 
> Hmm - says it's a network error? does hitting `https://index.crates.io` work in a browser?
> 
> Edit: also CI passed so i think it's probably fine


Running the `cargo test` on the main branch also results in an error, maybe it's my local network cause this error.

**Review by camc314 — APPROVED** [review]
Thank you!

**Pull request merged**